beacon-chain/rpc/lookup/blocker.go

```diff
@@ -212,64 +216,191 @@ func (p *BeaconDbBlocker) Blobs(ctx context.Context, id string, indices []int) (
+	if roBlock.Slot() >= fuluForkSlot {
+		roBlock, err := blocks.NewROBlockWithRoot(roSignedBlock, root)
+		if err != nil {
+			return nil, &core.RpcError{Err: errors.Wrapf(err, "failed to create roBlock with root %#x", root), Reason: core.Internal}
+		}
+
+		return p.blobsFromStoredDataColumns(roBlock, indices, rootSlice)
+	}
+
+	return p.blobsFromStoredBlobs(indices, rootSlice, commitments)
+}
+
+// blobsFromStoredBlobs retrieves blob sidercars corresponding to `indices` and `root` from the store.
+// This function expects blob sidecars to be stored (aka. no data column sidecars).
+func (p *BeaconDbBlocker) blobsFromStoredBlobs(indices []int, rootBytes []byte, commitments [][]byte) ([]*blocks.VerifiedROBlob, *core.RpcError) {
+	summary := p.BlobStorage.Summary(bytesutil.ToBytes32(rootBytes))
+	maxBlobCount := summary.MaxBlobsForEpoch()
+
+	for _, index := range indices {
+		if uint64(index) >= maxBlobCount {
+			return nil, &core.RpcError{
+				Err:    fmt.Errorf("requested index %d is bigger than the maximum possible blob count %d", index, maxBlobCount),
+				Reason: core.BadRequest,
 			}
 		}
-	} else {
-		for _, ix := range indices {
-			if uint64(ix) >= sum.MaxBlobsForEpoch() {
-				return nil, &core.RpcError{
-					Err:    fmt.Errorf("requested index %d is bigger than the maximum possible blob count %d", ix, sum.MaxBlobsForEpoch()),
-					Reason: core.BadRequest,
-				}
+
+		if !summary.HasIndex(uint64(index)) {
+			return nil, &core.RpcError{
+				Err:    fmt.Errorf("requested index %d not found", index),
+				Reason: core.NotFound,
 			}
-			if !sum.HasIndex(uint64(ix)) {
-				return nil, &core.RpcError{
-					Err:    fmt.Errorf("requested index %d not found", ix),
-					Reason: core.NotFound,
-				}
+		}
+	}
+
+	// If no indices are provided, use all indices that are available in the summary.
+	if len(indices) == 0 {
+		for index := range commitments {
+			if summary.HasIndex(uint64(index)) {
+				indices = append(indices, index)
 			}
 		}
 	}
 
-	blobs := make([]*blocks.VerifiedROBlob, len(indices))
-	for i, index := range indices {
-		vblob, err := p.BlobStorage.Get(root, uint64(index))
+	// Retrieve blob sidecars from the store.
+	root := bytesutil.ToBytes32(rootBytes)
+	blobs := make([]*blocks.VerifiedROBlob, 0, len(indices))
+	for _, index := range indices {
+		blobSidecar, err := p.BlobStorage.Get(root, uint64(index))
 		if err != nil {
 			return nil, &core.RpcError{
-				Err:    fmt.Errorf("could not retrieve blob for block root %#x at index %d", rootSlice, index),
+				Err:    fmt.Errorf("could not retrieve blob for block root %#x at index %d", root, index),
 				Reason: core.Internal,
 			}
 		}
-		blobs[i] = &vblob
+
+		blobs = append(blobs, &blobSidecar)
 	}
 
 	return blobs, nil
 }
+
+// blobsFromStoredDataColumns retrieves data column sidecars from the store,
+// reconstruct the whole matrix if needed, convert the matrix to blobs,
+// and then returns blobs corresponding to `indices` and `root` from the store.
```

Comment:
```suggestion
// blobsFromStoredDataColumns retrieves data column sidecars from the store,
// reconstructs the whole matrix if needed, converts the matrix to blobs,
// and then returns blobs corresponding to `indices` and `root` from the store.
```
But "returns blobs corresponding to `indices` and `root` from the store" sounds as if the blobs are in the store too. Would "return these blobs" be more precise?

Reply:
"return these blobs"  would mean "return all reconstructed" blobs. Let me try something.

Reply:
What about b84697ed188fc8b1442154681a352a7e449c54b0?